universal-application-tool-0.0.1/app/repository/QuestionRepository.java

```diff
@@ -33,6 +44,56 @@ public QuestionRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext exec
         () -> ebeanServer.find(Question.class).setId(id).findOneOrEmpty(), executionContext);
   }
 
+  public Question getOrCreateDraft(QuestionDefinition definition) {
+    try {
+      Transaction transaction = ebeanServer.beginTransaction();
+      Optional<Question> draftMaybe =
+          ebeanServer
+              .find(Question.class)
+              .usingTransaction(transaction)
+              .where()
+              .eq("name", definition.getName())
+              .eq("lifecycle_stage", LifecycleStage.DRAFT.getValue())
+              .findOneOrEmpty();
+      if (draftMaybe.isPresent()) {
+        return draftMaybe.get();
+      }
+      int existingQuestionCount =
+          ebeanServer
+              .find(Question.class)
+              .usingTransaction(transaction)
+              .where()
+              .eq("name", definition.getName())
+              .findCount();
+      Question newDraft =
+          new Question(
+              new QuestionDefinitionBuilder(definition)
+                  .setId(null)
+                  .setVersion(existingQuestionCount + 1L)
+                  .build());
+      newDraft.setLifecycleStage(LifecycleStage.DRAFT);
+      newDraft.id = null;
+      ebeanServer.insert(newDraft, transaction);
+      for (Program draftProgram :
+          ebeanServer
+              .find(Program.class)
+              .where()
+              .eq("lifecycle_stage", LifecycleStage.DRAFT.getValue())
+              .findList()) {
+        programRepository.updateQuestionVersions(draftProgram, transaction);
+      }
+      ebeanServer.commitTransaction();
+      newDraft.refresh();
+      return newDraft;
+    } catch (UnsupportedQuestionTypeException e) {
+      // The question already exists - it should not be unsupported.  Wrap with runtime exception
+      // so the crash is maintained but callers do not have to deal with this.
+      throw new RuntimeException(e);
+    } finally {
+      ebeanServer.endTransaction();
+    }
+  }
+
```

Comment:
Suggestion (non-blocking): I would find a lot of this could easier to read with empty lines between logically distinct portions

Reply:
thanks!